### ISSUE TITLE
installing missing theme snaps failed

### ISSUE THREAD

**Issue by fossfreedom** [canonical/snapd-desktop-integration#52]
On clicking our notification to install our theme snap on Lunar we receive an error notification saying the installation failed.

This is odd because looking at /snap we can see that our theme snap gtk-theme-pocillo has been downloaded and installed.

looking at journalctl I see this
![image](https://user-images.githubusercontent.com/996240/228350030-36af41bb-7f94-4be5-a555-d978f0eaa6de.png)

Ubuntu Budgie 23.04 

Our default theme is Pocillo-dark.

**Comment by fossfreedom** [comment]
Further to this (v0.9 snapd-desktop-integration installed),  clicking No to the notification actually falls through to try and install the theme snap.

https://github.com/snapcore/snapd-desktop-integration/blob/5960b4ae320d2d0a182b001d837016f7294db5f5/src/sdi-theme-monitor.c#L154 here a "default" is added to the yes/no answers and calling notify-cb - 

so here https://github.com/snapcore/snapd-desktop-integration/blob/5960b4ae320d2d0a182b001d837016f7294db5f5/src/sdi-theme-monitor.c#L101 its trying to match either "yes" or "default" so is a  "no" being translated as a "default" answer and thus the install routine being called?

**Comment by kred-jon** [comment]
fixed or not? because i received it the same problem

**Comment by Mr-McGregor** [comment]
It's still not fixed on 25.04.

**Comment by sergio-costas** [comment]
Can you check that all the interfaces of snapd-desktop-integration are connected?

`snap connections snapd-desktop-integration`

**Comment by Mr-McGregor** [comment]
Thanks for your answer, Sergio!
UB 24.04

`snap connections snapd-desktop-integration

Interface               Connecteur                                       Prise                                                Notes
content[gnome-42-2204]  snapd-desktop-integration:gnome-42-2204          gnome-42-2204:gnome-42-2204                          -
content[gtk-3-themes]   snapd-desktop-integration:gtk-3-themes           gtk-common-themes:gtk-3-themes                       -
content[gtk-3-themes]   snapd-desktop-integration:gtk-3-themes           gtk-theme-pocillo:gtk-3-themes                       -
content[icon-themes]    snapd-desktop-integration:icon-themes            gtk-common-themes:icon-themes                        -
content[icon-themes]    snapd-desktop-integration:icon-themes            icon-theme-papirus:icon-themes                       -
content[sound-themes]   snapd-desktop-integration:sound-themes           gtk-common-themes:sound-themes                       -
dbus                    -                                                snapd-desktop-integration:snapd-desktop-integration  -
desktop                 snapd-desktop-integration:desktop                :desktop                                             -
desktop-launch          snapd-desktop-integration:desktop-launch         :desktop-launch                                      -
desktop-legacy          snapd-desktop-integration:desktop-legacy         :desktop-legacy                                      -
gsettings               snapd-desktop-integration:gsettings              :gsettings                                           -
login-session-observe   snapd-desktop-integration:login-session-observe  :login-session-observe                               -
opengl                  snapd-desktop-integration:opengl                 :opengl                                              -
snap-refresh-observe    snapd-desktop-integration:snap-refresh-observe   :snap-refresh-observe                                -
snap-themes-control     snapd-desktop-integration:snap-themes-control    :snap-themes-control                                 -
system-files            snapd-desktop-integration:snap                   :system-files                                        -
unity7                  snapd-desktop-integration:unity7                 :unity7                                              -
wayland                 snapd-desktop-integration:wayland                :wayland                                             -
x11                     snapd-desktop-integration:x11                    :x11              `

**Comment by sergio-costas** [comment]
Yes, seems that everything is connected...

**Comment by Mr-McGregor** [comment]
I'm not blaming anyone, but I am surprised that a problem of this kind has existed for three years and is repeated from version to version.
It's not good for UB's image.

**Comment by kred-jon** [comment]
The integration seems to continuously attempt to update the missing theme under Snap, even though no theme is available. It keeps trying to match with the GTK theme.

It should first check if an update is available, and only show the update prompt if one is present. If no update is available, it should ignore the update notification.

**Comment by sergio-costas** [comment]
BTW: the `default` option is when the user clicks on the notification, instead of clicking on "Yes" or "No". That's why we compare with those two values to decide whether we should install or not the snap with the theme.

**Comment by fossfreedom** [comment]
@sergio-costas much appreciated for looking into this.  I raised this issue around issues when you hit the "no" button in the notification.  The expectation was that the snap would not be installed.

**Comment by sergio-costas** [comment]
Mmm... I'll test it, then... odd...

**Comment by sergio-costas** [comment]
Well, this is odd... I switched to `Yaru`, removed `gtk-theme-pocillo`, switched again to `Pocillo-dark`, a notification asking to install `Pocillo theme for snaps` appeared as expected, I clicked on `No`, and nothing was installed... I switched again to `Yaru`, then again to `Pocillo-dark`, the notification appeared again, I clicked over the notification (in the text itself, neither over the `Yes`, nor over the `No`), and it installed the theme as expected...

**Comment by sergio-costas** [comment]
Also tested clicking on the "X" to specifically close the notification, and again, the theme wasn't installed...

**Comment by kred-jon** [comment]
Currently, on every boot, a prompt appears requesting the installation of a theme without verifying its availability in the Snap repository. 

I'm using "New theme: gtk=Qogir-Ubuntu-Dark icon=Qogir-ubuntu cursor=Breeze-Snow, sound=Smooth"

**Comment by kred-jon** [comment]
hmmm somehow is good today:

2025-04-09T08:45:44.862036+08:00 noble snapd-desktop-i[2197]: New theme: gtk=Default icon=Adwaita cursor=(null), sound=freedesktop
2025-04-09T08:45:45.429422+08:00 noble snapd-desktop-i[2197]: All available theme snaps installed
2025-04-09T08:45:45.808129+08:00 noble snapd-desktop-i[2197]: New theme: gtk=Qogir-Ubuntu icon=Qogir-ubuntu cursor=Qogir-ubuntu-dark, sound=Smooth
2025-04-09T08:45:50.327055+08:00 noble snapd-desktop-i[2197]: All available theme snaps installed

**Comment by Mr-McGregor** [comment]
To summarise — and clarify if possible — the system offers to install the snaps packages required for the gtk theme you want to use (for example, pocillo slim, dark or light):

![Image](https://github.com/user-attachments/assets/407ad081-ce24-4732-8c22-d93bd05dbed3)

- If you click "Yes", a failure message will be displayed... even though the snap package has been installed:

![Image](https://github.com/user-attachments/assets/b576cfc9-91e5-4be0-bcd5-48cea231a166)

`snap list
Nom                        Version          Révision  Suivi            Éditeur       Notes
gtk-common-themes          0.1-81-g442e511  1535      latest/stable/…  canonical✓    -
gtk-theme-pocillo          0.21.0           7         latest/stable    ubuntubudgie  -`

If you click "No", a success message is displayed...

![Image](https://github.com/user-attachments/assets/80c1df7b-1af1-4ff6-ba11-76dc03136c24)

One wonders why these snap packages are not included in the installation — Firefox is, although we'd prefer another version...

And this has been the case since UB 24.04. I'm sure you'll agree that it's a mess --- and there are other "details" of the same kind.

Sorry for the French screenshots: UB 25.04 has gone backwards on this point: you can install a second language, but you can no longer activate it...

**Comment by sergio-costas** [comment]
Mmmm... wait... so, if you say "No" to the installation of the snap, it says that it has been correctly installed...?

And if you say "Yes", it returns an error, but, at the same time, it says that everything was correctly installed...?

Odd... Maybe only part of the snaps have been installed? Remember that each change shows a notification (if you change the icons, then the cursors, then the theme, you will have one notification per change... maybe is that what is happening?).

**Comment by sergio-costas** [comment]
Anyway, it certainly should be grouped, or shown specifically to WHAT refers each notification... I'll do some tests and discuss this with Design.

**Comment by Mr-McGregor** [comment]
Thanks Sergio.
What I described is easy to verify from a fresh BU 25.04 installation by checking the result with `snap list`.
These oddities have been going on since UB 24.04 and are well known on the various internet forums.

**Comment by sergio-costas** [comment]
I've been trying to reproduce these errors, but no dice. The only thing that looked like a possibility was a race condition, if you change one thing (let's say, the gtk theme), say "yes" to install it, and while installing, you change a different thing (let's say, the icon theme). But even in that case it doesn't seem to fail...

**Comment by sergio-costas** [comment]
Can you tell me ALL your configuration? Not only gtk theme but also icons theme and cursor theme, to try to reproduce this.

**Comment by sergio-costas** [comment]
(I presume that you are using gnome-tweaks application...)

**Comment by sergio-costas** [comment]
@Mr-McGregor Can you point me to those forums, to try to gather more info about this problem? As I said, I can't reproduce it (with the obvious exception of the asking to install it after every boot...)

**Comment by fossfreedom** [comment]
@sergio-costas gnome-tweaks?  ok - sounds like you are on gnome-shell.  That could be a factor.  Ubuntu Budgie is was we are using  - if you do have the bandwidth and a spare virtual, then from the menu choose budgie settings and you can change the theme there

**Comment by Mr-McGregor** [comment]
@sergio-costas Yes, this is Ubuntu Budgie, and no need for “gnome-tweaks” for that.

I can't describe the problem better than above.
I have installed the Orchis theme on Ubuntu Budgie 24.04 and the same message appears every time I log on:

![Image](https://github.com/user-attachments/assets/bbe0ee7c-17da-444a-8798-20d37799ca69)

However, 'gtk-theme-orchis” is installed:

`snap list
gtk-common-themes                     0.1-81-g442e511                 1535   latest/stable    canonical✓        -
gtk-theme-orchis                      2022-07-20                      3      latest/stable    eeickmeyer        -
gtk-theme-pocillo                     0.21.0                          7      latest/stable    ubuntubudgie      -
`

Both “Yes” and “No” buttons install the snap, but the “Yes” button returns a failure message while the “No” button returns a success message.

And the latter problem continues with UB 24.10 and 25.04.

As for the forums I was referring to, a [Google search with these keywords](https://www.google.com/search?q=Ubuntu+Budgie+24.04+some+required+theme+snaps+are+missing) immediately yields results.

**Comment by sergio-costas** [comment]
Ok, don't worry. I'll try to install Ubundu Budgie in a virtual machine and try to reproduce there the problem.

**Comment by sergio-costas** [comment]
Ok, I installed it and found that when snapd-desktop-integration asks for installing the missing themes, there are two notifications, not one. That can explain the error 500: two updates are started at the same time, so one fails due to the conflict. Now I need to know why is happening this.

**Comment by Mr-McGregor** [comment]
Cool stuff, @sergio-costas!
Now we need to understand why “No” installs “successfully” when it shouldn't, and why “Yes” also installs, but “unsuccessfully”.
And why the message keeps triggering even though the theme snap is installed (but I think that's been fixed since UB 24.04).
Or, more simplt, why doesn't Ubuntu Budgie come natively with these gtk theme snaps installed, since it's stupid to ask if you want to install them anyway, since they seem necessary or at least useful — but Solus Budgie doesn't use them.

**Comment by sergio-costas** [comment]
I'm preparing an instrumented version of snapd-desktop-integration to see exactly what is being sent and what's being received. That should allow us to know what is happening.

**Comment by sergio-costas** [comment]
Ok, I found the problem: the notification manager of that desktop is sending wrong events. There are three events that a notification can send: "close" (when the notification is closed), "default" (when the user clicks over the notification, in the text), and a specific action name (when the user clicks on a button in a notification; for example, when pressing "yes" or "no" in the notification that asks whether to install or not the missing themes).

The problem is that, when the user clicks on a button, the notification manager sends the action for that button, followed by the "default" action; the correct behavior is to only send the button action, nothing else.

This explains the problems detected: since the "default" action installs the missing themes, clicking on "yes" sends two actions to install themes ("yes" and "default"), so the second one interferes with the first one, and that's why an error is shown, even when the theme is correctly installed. Also, pressing on "no" sends one action for "do nothing" ("no") and another action to install the themes ("default"), and that's why it installs it anyway.

In Gnome Shell and in Kde Plasma, the behavior is the expected one: only one action.

This little program shows the bug:

```
#!/usr/bin/env python3

import notify2
import dbus
import dbus.mainloop.glib
from gi.repository import GLib

loop = GLib.MainLoop()

dbus_mainloop = dbus.mainloop.glib.DBusGMainLoop()
notify2.init("test-notify", dbus_mainloop)

def cb(notification, action):
    print(f"Received action: {action}")

def closedcb(notification):
    print(f"Received CLOSE action")
    loop.quit()

n=notify2.Notification("A test notification", "Do click on the buttons, the close button, or the notification itself.", "")
n.add_action("yes", "Yes", cb)
n.add_action("no", "No", cb)
n.add_action("default", "Default", cb)
n.connect("closed", closedcb)
n.show()

loop.run()
```

**Comment by fossfreedom** [comment]
Really appreciate narrowing this down - this will most definitely help finally resolving this once and for all.

Raised this upstream https://github.com/BuddiesOfBudgie/budgie-desktop/issues/700

I'll close this and manage this upstream.

**Comment by sergio-costas** [comment]
Thanks for raising this to upstream. Anyway, I'll add a workaround in snapd-desktop-integration, to ensure that this doesn't happen again.

**Comment by mustaqimM** [comment]
@sergio-costas I get this on the Gnome desktop too; I don't think it's solely a Budgie issue. 

For example, toggling the Style from Default to Dark, on every startup, or running `/snap/snapd-desktop-integration/253/usr/bin/snapd-desktop-integration` will show a popup but it always fails for the past 3 versions of Ubuntu:

> Some required theme snaps are missing. Would you like to install them now? I click Yes on the prompt, enter my password but then a few seconds later, another notification says: Installing missing theme snaps: Failed.

![Image](https://github.com/user-attachments/assets/e00b2094-6f92-4a9c-81cb-a4ea0f9bb116)


`journalctl -b0 > journal.txt`:
```bash
May 02 05:27:31 ubuntu snapd-desktop-i[155708]: Missing theme snaps
May 02 05:27:35 ubuntu dbus-daemon[3250]: apparmor="DENIED" operation="dbus_signal"  bus="session" path="/org/freedesktop/Notifications" interface="org.freedesktop.Notifications" member="ActivationToken" name=":1.40" mask="receive" pid=155708 label="snap.snapd-desktop-integration.snapd-desktop-integration" peer_pid=3524 peer_label="unconfined"
May 02 05:27:35 ubuntu snapd-desktop-i[155708]: Installing missing theme snaps...
May 02 05:27:36 ubuntu snapd-desktop-i[155708]: Installation failed: status-code=500 kind=(null) message=cannot install themes: snap "icon-theme-breeze" is already installed
```

```zsh
$ snap connections snapd-desktop-integration
Interface               Plug                                             Slot                                                 Notes
content[gnome-42-2204]  snapd-desktop-integration:gnome-42-2204          gnome-42-2204:gnome-42-2204                          -
content[gtk-3-themes]   snapd-desktop-integration:gtk-3-themes           gtk-common-themes:gtk-3-themes                       -
content[icon-themes]    snapd-desktop-integration:icon-themes            gtk-common-themes:icon-themes                        -
content[sound-themes]   snapd-desktop-integration:sound-themes           gtk-common-themes:sound-themes                       -
dbus                    -                                                snapd-desktop-integration:snapd-desktop-integration  -
desktop                 snapd-desktop-integration:desktop                :desktop                                             -
desktop-launch          snapd-desktop-integration:desktop-launch         :desktop-launch                                      -
desktop-legacy          snapd-desktop-integration:desktop-legacy         :desktop-legacy                                      -
gsettings               snapd-desktop-integration:gsettings              :gsettings                                           -
login-session-observe   snapd-desktop-integration:login-session-observe  :login-session-observe                               -
opengl                  snapd-desktop-integration:opengl                 :opengl                                              -
snap-refresh-observe    snapd-desktop-integration:snap-refresh-observe   :snap-refresh-observe                                -
snap-themes-control     snapd-desktop-integration:snap-themes-control    :snap-themes-control                                 -
system-files            snapd-desktop-integration:snap                   :system-files                                        -
unity7                  snapd-desktop-integration:unity7                 :unity7                                              -
wayland                 snapd-desktop-integration:wayland                :wayland                                             -
x11                     snapd-desktop-integration:x11                    :x11                                                 -
```

**Comment by sergio-costas** [comment]
I presume that you are using `breeze`... isn't it? Can you send me all your installed `gtk-theme-XXXX` snaps, and the specific theme you are using?

**Comment by mustaqimM** [comment]
> I presume that you are using `breeze`... isn't it?

No, I'm not; I'm on Gnome on the default Ubuntu Yaru theme. In the past I installed KDE but I don't have any additional themes installed, including Breeze or what it's calling "icon-theme-breeze" (I don't see it anywhere)

**Comment by sergio-costas** [comment]
Mmm... ok, that gives me a thread to pull... anyway, can you open a different bug for this, please?
